Added prisma, defined models

ccweaver1 opened the pull request:
Started down the path of integrating Prisma as an ORM and defining database models that are scalable and allow us to expand this platform beyond a single league/season.

• Added prisma
• Laid out a fairly comprehensive DB schema that allows for multiple leagues, multiple seasons within those leagues, and association between players and league-seasons
• Made some changes to existing models (players, teams, games) where normalization or stronger typing made sense to me

Open to feedback and discussion!

apathetic commented on prisma/schema.prisma:
```diff
@@ -0,0 +1,159 @@
+// So if in '22 I played on CK in BTSH and Team Blue in D5 then I would have 2 entries
+// This allows pulling all players in a league-season
+// And also pulling a team's roster for a league-season
+model PlayerLeagueSeasonLink {
```
this name is a little convoluted to me 😃   is it essentially the `TeamRoster` ?

jmichaliga replied:
does this also account for free-agents too? If so, that would be pretty interesting.

ccweaver1 replied:
if we made teamSlug optional, or added a `Team` for free agents then yes this could work for free agents as well.

I also thought about calling this a `TeamRoster` but in my experience using some sort of convention for these many-to-many linking tables helps to keep your code more understandable down the line.

apathetic commented on prisma/schema.prisma:
```diff
@@ -0,0 +1,159 @@
+}
+
+// An instance of a league in a season ("btsh-2022")
+model LeagueSeason {
```
i'd also vote for just `Season` for simplicity; i think it's obvious it's per-league with the @relations / nice comment 🤷

apathetic commented on prisma/schema.prisma:
```diff
@@ -0,0 +1,159 @@
+    lastname  String
+    position  String?
+    gender    String?
+    // email?
+    // phone?
+    // birthdate?
+    // favoriteColor?
```
I think also 

name
display_name
email
photo?
team (if not Roster/leagueSeason)
accepted_terms}</Paragraph>
is_captain or maybe `roles` array

apathetic commented on prisma/schema.prisma:
```diff
@@ -0,0 +1,159 @@
+    description       String?
+    week              Int
+    status            String   @default("SCHEDULED")
+    homeScore         Int
+    awayScore         Int
+    homeScoreByPeriod Int[]
+    awayScoreByPeriod Int[]
```
is there any concern that having both these is potentially error prone/could lead to data drift?   it could be nice to have a "single source of truth".  e.g it's trivial to get homeScore by summing the array

jmichaliga replied:
I think this is nice, we may not have all of the period info for past games. Thinking for backwards compatibility as well.

apathetic commented on prisma/schema.prisma:
```diff
@@ -0,0 +1,159 @@
+    // ^ Prisma complains if this isn't here, but I'm not sure if its a backref
+    // that will ever get any use
+
+    // How do we handle bye weeks and holidays?
```
i'm not sure i follow... why should `Game` care?  wouldn't the datetime be sufficient ?   like, if this just the model to hold the game data, we shouldnt conflate scheduling

apathetic commented on prisma/schema.prisma:
```diff
@@ -0,0 +1,159 @@
+    awayTeamSlug      String
+    title             String?
+    description       String?
+    week              Int
```
per my comment below - i think this here is a little odd, and scheduling should be kept separate

jmichaliga replied:
agree we probably could take out week. I may have had that in an earlier model. It ended up not being too relevant when games were rained out or had to be rescheduled. this is likely just a BTSH thing.

apathetic commented on prisma/schema.prisma:
```diff
@@ -0,0 +1,159 @@
+    court             String
+    homeTeamSlug      String
+    awayTeamSlug      String
+    title             String?
+    description       String?
```
nit: prob don't need

apathetic reviewed (approved):
dude, nice! this is great!

ccweaver1 commented:
Thanks for the feedback Wes, definitely agree with you on some of those notes.  Just waiting to have some time to actually respond to each and push updates.

jmichaliga reviewed (approved):
Leaving some of these for more review, but as far as getting things started and the ball rolling, I'd say it's good to start from.

ccweaver1 commented on prisma/schema.prisma:
```diff
@@ -0,0 +1,159 @@
+
+    leagueSeasons   PlayerLeagueSeasonLink[]
+    playerEvents    Event[]                  @relation("Player")
+    oppGoalieEvents Event[]                  @relation("OppGoalie")
```
this is mostly there because Prisma doesn't like if you don't declare the backref.  I'm not sure it will ever be useful

ccweaver1 commented:
- Re-added a "Stats" tab to the layout in order to test some SSR prisma queries against new events table.
- Moved old stats to a separate "hardcodedstats" file
- Wrote raw SQL queries parameterized by prisma for join/aggregation query on events.  This will allow for future filtering on more complex attr. (e.g. player position, gender.  could also join teams, split by division, etc)
- Wrote pivot function to transform result of aggregated events into rows of player stats
